### ISSUE TITLE
Protobuf for regular JSON only?

### ISSUE THREAD

**Issue by Julio-Guerra** [gogo/protobuf#541]
Hello,

Does it make sense if I tell you that I use proto files to specifiy my REST API, generate all the structs by making benefit from protoc plugins such as gogoproto? Which means I don't need all the PB marshaling methods and helper...

If it does, I am willing to add or modify gogoproto to only generate regular json.

Thanks for your feedback!

**Comment by jmarais** [comment]
Hi.

> Does it make sense if I tell you that I use proto files to specifiy my REST API, generate all the structs by making benefit from protoc plugins such as gogoproto?

Are you saying you define your API with the ```.proto``` files and then generating it using the gogoproto protoc plugin?
That sounds fine. You might even want to use GRPC. Here is an example:
https://github.com/gogo/grpc-example

>  Which means I don't need all the PB marshaling methods and helper...
> If it does, I am willing to add or modify gogoproto to only generate regular json.

I am guessing you are then just un/marshalling to and from json?
You can do that without using the un/marshaling options. You can just use the ```"github.com/gogo/protobuf/jsonpb"``` package to marshal to and from json:

```golang
	msg := &Foo{
		Bar: "Hello",
	}
	marshaler := &jsonpb.Marshaler{}
	str, err := marshaler.MarshalToString(msg)
	if err != nil {
		panic(err)
	}
	result := &Foo{}
	err = jsonpb.Unmarshal(strings.NewReader(str), result)
	if err != nil {
		panic(err)
	}
```

> If it does, I am willing to add or modify gogoproto to only generate regular json.

Does the above example help you?

**Comment by Julio-Guerra** [comment]
> I am guessing you are then just un/marshalling to and from json?
> You can do that without using the un/marshaling options. You can just use the `"github.com/gogo/protobuf/jsonpb"` package to marshal to and from json:
> 
> ```go
> 	msg := &Foo{
> 		Bar: "Hello",
> 	}
> 	marshaler := &jsonpb.Marshaler{}
> 	str, err := marshaler.MarshalToString(msg)
> 	if err != nil {
> 		panic(err)
> 	}
> 	result := &Foo{}
> 	err = jsonpb.Unmarshal(strings.NewReader(str), result)
> 	if err != nil {
> 		panic(err)
> 	}
> ```
> 
> > If it does, I am willing to add or modify gogoproto to only generate regular json.
> 
> Does the above example help you?

That's basically my original source of confusion: `jsonpb` allows to marshal pb to json, and conversely. And this is misleading when you start because protobufs have more "type information" than regular json, so marshaling protobufs to json don't give you a regular json but rather a json + extra fields for context and metadata.

So for example, marshaling to json a protobuf `Struct` gives the struct's internal fields and not a straight json struct.

```
{
  "fields": {
     "f1": ...,
     "f2": ...
  }
}
```

instead of:

```
{
   "f1": ...,
   "f2": ...
}
```

So basically, I stopped using `jsonpb` and switched back to `json` + custom `MarshalJSON()` methods when required (so far for `oneof`, `Struct` and `ListValue`).

So if you think this is a valid use-case, I am willing to add the support for it (by having another generator for this json-only purpose, and generating `MarshalJSON()` methods, and asking for user-provided `UnmarshalJSON()` implementations when necessary).

**Comment by jmarais** [comment]
Ah ok. I think I understand your problem. However, I do not think I can recreate it. I created an example here using ```oneoff``` and ```ListValue```:
https://github.com/jmarais/protobuf/tree/issue541/test/issue541
protofile:
```protobuf
syntax = "proto3";

package issue541;

import "github.com/gogo/protobuf/gogoproto/gogo.proto";
import "google/protobuf/struct.proto";

message OuterMessage {
	string fieldone = 1;
	InnerMessage inner = 2;
	oneof OneOfFoo {
		int64 fooOne = 3;
		int32 fooTwo = 4;
	}
	google.protobuf.ListValue BazListValue = 5;
}

message InnerMessage {
	string fieldtwo = 1;
	string fieldthree = 2;
}
```
When I then just do:
```golang
marshaler := &jsonpb.Marshaler{}
str, _ := marshaler.MarshalToString(msg)
fmt.Printf("%s", str)
```
I get the following json output:
```json
{
	"fieldone":"Hello",
	"inner":{
		"fieldtwo":"Sailboat",
		"fieldthree":"Sailor"
	},
	"fooOne":"2",
	"BazListValue":[
		"ListValueValueOne"
	]
}
```

Could you provide a sample .proto and golang example struct which highlights your issue with the jsonpb marshaller?

**Comment by Julio-Guerra** [comment]
Sorry about the delay.

I isolated my original test into the following test function which indeed passes as expected. So I no longer can tell what happened at the time it failed (maybe I used `encoding/json` instead of jsonpb :sweat: )...

```
func TestStruct(t *testing.T) {
	pb := &types.Struct{
		Fields: map[string]*types.Value{
			"field": &types.Value{
				Kind: &types.Value_StringValue{"a string"},
			},
		},
	}

	// Check it can be marshaled to the expected JSON struct.
	marshaler := &jsonpb.Marshaler{}
	str, err := marshaler.MarshalToString(pb)
	require.NoError(t, err)
	require.Equal(t, str, `{"field":"a string"}`)

	// Check it can be unmarshaled back to Protobuf.
	parsedPB := new(types.Struct)
	err = jsonpb.UnmarshalString(str, parsedPB)
	require.NoError(t, err)
	require.Equal(t, parsedPB, pb)
}
```

Last related opened questions:

- How can we make `Struct` and `ListValue` easier to use by being closer to Go representations `interface{}` and `[]interface{}`? It would avoid transforming/copying Go values into `Struct` and `ListValue` temporary variables, right?

- How can I disable all protobuf marshaler/unmarshaler methods from the generated code to avoid having code I don't use when only marshaling/unmarshaling JSON (ie. code I don't cover nor test because pointless to my usage)?
